src/network/network.cpp

```diff
@@ -1424,9 +1462,49 @@ void Network::Server_Handle_TOKEN(NetworkConnection& connection, NetworkPacket&
 	for (int i = 0; i < token_size; i++) {
 		connection.Challenge[i] = (uint8)(rand() & 0xff);
 	}
+	rct_object_entry object_entries[OBJECT_ENTRY_COUNT];
+	int count = scenario_get_num_packed_objects_to_write(object_entries);
+	Server_Send_OBJECTS(connection, object_entries, count);
 	Server_Send_TOKEN(connection);
 }
 
+void Network::Client_Handle_OBJECTS(NetworkConnection& connection, NetworkPacket& packet)
+{
+	IObjectRepository * repo = GetObjectRepository();
+	uint32 size;
+	packet >> size;
+	log_verbose("client received object list, it has %u entries", size);
+	std::vector<std::string> requested_objects;
+	for (uint32 i = 0; i < size; i++)
+	{
+		const char * name = (const char *)packet.Read(8);
+		uint32 checksum;
+		packet >> checksum;
+		std::string s(name, name + 8);
+		const ObjectRepositoryItem * ori = repo->FindObject(s.c_str());
+		if (ori == nullptr || ori->ObjectEntry.checksum != checksum) {
+			log_verbose("Requesting object %s with checksum %x from server",
+						s.c_str(), checksum);
+			requested_objects.push_back(s);
+		}
+	}
+	Client_Send_OBJECTS(requested_objects);
+}
+
+void Network::Server_Handle_OBJECTS(NetworkConnection& connection, NetworkPacket& packet)
+{
+	uint32 size;
+	packet >> size;
+	log_verbose("Client requested %u objects", size);
+	for (uint32 i = 0; i < size; i++)
+	{
+		const char * name = (const char *)packet.Read(8);
+		std::string s(name, name + 8);
+		log_verbose("Client requested object %s", s.c_str());
+		connection.RequestedObjects.push_back(s);
+	}
```

Comment:
You need to check if an object is custom or not. Otherwise a client could request vanilla / expansion objects and receive them. The server should refuse this.